source/extensions/transport_sockets/http_11_proxy/connect.cc

```diff
@@ -57,8 +70,22 @@ UpstreamHttp11ConnectSocket::UpstreamHttp11ConnectSocket(
     const bool has_proxy_addr = metadata->typed_filter_metadata().contains(
         Config::MetadataFilters::get().ENVOY_HTTP11_PROXY_TRANSPORT_SOCKET_ADDR);
     if (has_proxy_addr) {
-      header_buffer_.add(
-          absl::StrCat("CONNECT ", host->address()->asStringView(), " HTTP/1.1\r\n\r\n"));
+      if (!Runtime::runtimeFeatureEnabled(
+              "envoy.reloadable_features.http_11_proxy_connect_legacy_format")) {
+        // Prefer <host-name>:<port> for RFC 9110 compliance, unless URI is <host-ip>:<port>.
+        std::string target;
+        if (!host->hostname().empty()) {
+          const uint32_t port = host->address()->ip()->port();
+          target = absl::StrCat(host->hostname(), ":", port);
+        } else {
+          target = host->address()->asStringView();
+        }
+        header_buffer_.add(formatConnectRequest(target));
+      } else {
+        // Legacy behavior: <host-ip>:<port> format, no Host header for backward compatibility.
+        header_buffer_.add(
+            absl::StrCat("CONNECT ", host->address()->asStringView(), " HTTP/1.1\r\n\r\n"));
+      }
```

Comment:
Can you address this nesting by either refactoring or breaking some of the logic into another function? It's just too deep.

Reply:
Done, PTAL.
Thanks again for the guidance.